src/system/SystemPacketBuffer.cpp

```diff
@@ -634,12 +634,21 @@ PacketBufferHandle PacketBufferHandle::New(size_t aAvailableSize, uint16_t aRese
     SYSTEM_STATS_INCREMENT(chip::System::Stats::kSystemLayer_NumPacketBufs);
 
     lPacket->payload = lPacket->ReserveStart() + aReservedSize;
-    lPacket->len = lPacket->tot_len = 0;
-    lPacket->next                   = nullptr;
-    lPacket->ref                    = 1;
 #if CHIP_SYSTEM_PACKETBUFFER_FROM_CHIP_HEAP
     lPacket->alloc_size = lAllocSize;
 #endif
+#if !CHIP_SYSTEM_CONFIG_USE_LWIP
+    // For non-lwip packet buffer, the allocated packet buffer will not have chained buffers, set it to null to avoid potential
+    // issues.
+    lPacket->next = nullptr;
+#endif
+    // Set ther length and total length of the head packet buffer and all the chained packet buffers to 0
+    // as we don't put any data in them. And set the ref to 1 for all the buffers.
+    for (PacketBuffer * pktBuf = lPacket; pktBuf != nullptr; pktBuf = pktBuf->ChainedBuffer())
```

Comment:
So...  I am quite sure that various callers of `PacketBufferHandle::New` expect to get back a non-chained buffer and will fail if that is not the case.  Other failure modes just in this function, if for some reason we have a chained buffer: what if aReservedSize is larger than the size of the head buffer?  Then we have a corrupt `lPacket->payload`...

I would suggest splitting this piece out of the TCP changes, since it's not directly related, so we can figure out what should actually happen here.

Reply:
For the platforms which don't use lwip, the allocated packet buffer will never have a chained buffer, but for platforms using LwIP, the `pbuf_alloc()` will allocate pbuf with chained pbufs if the allocated size is too large, and in our current code the chained pbufs will never be released which causes memory leak.

Agreed that we should split the change, reverted it in this PR.

Reply:
My point is not that we don't need this change, but rather that we probably need a complete rethink of how we use packetbuffers if `New` can ever create a chained buffer.  It will break crypto code, it will break all sorts of stuff....